### PR TITLE
Enabling single precision + deletion of virtual status of functions + style fixes + logs

### PR DESCRIPTION
**Pull request by GiudGiud**
A lot of little things in this PR

This PR enables running 3D BEAVRS assemblies in linear source with single floating point precision, by simply setting the segment lengths to be doubles all the time.
This improves the accuracy of the centroids, which helps converging the linear source.

By making the volumes double precision (not done here), the centroid accuracy is further improved, so if anyone runs into issues with single precision and (linear source) that disappear with double precision, this is a place to start at.

### CODE REVIEW

File: src/Progress.cpp [hunk 1]
```diff
@@ -57,11 +70,13 @@ void Progress::incrementCounter() {
       if (_curr_interval >= _intervals.size())
         break;
     }
-#pragma omp flush (_counter, _curr_interval)
```
Comment (GiudGiud) [inline src/Progress.cpp]: this flush ought to be deleted, since flushes are implied at the end of critical sections

File: src/Region.cpp [hunk 2]
```diff
@@ -313,7 +313,7 @@ double Region::getMaxZ() {
   else if(_region_type == UNION)
     max_z = -std::numeric_limits<double>::infinity();
   else if(_region_type == COMPLEMENT) {
-    log_printf(WARNING, "getMaxZ() is not implemented for complement regions,"
+    log_printf(DEBUG, "getMaxZ() is not implemented for complement regions,"
```
Comment (GiudGiud) [inline src/Region.cpp]: this log was too annoying, as it is triggered for every 3D geometry

File: src/Universe.cpp [hunk 3]
```diff
@@ -1709,8 +1711,8 @@ Cell* Lattice::findCell(LocalCoords* coords) {
   double next_z = coords->getZ()
                   - (getMinZ() + _widths_z[lat_z]/2. + _accumulate_z[lat_z]);
 
-  /* Check for 2D problem */
-  if (_width_z == std::numeric_limits<double>::infinity())
+  /* Check for 2D problem or 2D lattice */
+  if (_width_z > FLT_INFINITY)
```
Comment (GiudGiud) [inline src/Universe.cpp]: I ran into the case that the line above was not working. I think it's very much bad luck since the double is supposed to be set to this value exactly. 
Either way, I think this is a good substitute and we should think about using it in more places.

File: profile/Makefile [hunk 4]
```diff
@@ -182,10 +183,10 @@ CFLAGS += -DVEC_ALIGNMENT=16
 # Optimization Flags
 ifeq ($(OPTIMIZE),yes)
 ifeq ($(COMPILER),gnu)
-  CFLAGS += -O3 -ffast-math -fpic
+  CFLAGS += -O3 -ffast-math -fpic -march=native
```
Comment (GiudGiud) [inline profile/Makefile]: this flag turns on architecture dependent optimizations. Since everyone is building OpenMOC from source it's fine. 
If the code was "clear" enough to the compiler, this also turns on SIMD vectorization

SIMD will be part of another PR

File: src/TrackGenerator3D.cpp [hunk 5]
```diff
@@ -1216,13 +1216,28 @@ void TrackGenerator3D::segmentizeExtruded() {
...
+#ifdef MPIx
+  if (_geometry->isDomainDecomposed()) {
+    MPI_Allreduce(&num_Z_planes, &_num_Z_planes, 1, MPI_INT, MPI_SUM,
+                  _geometry->getMPICart());
+    MPI_Allreduce(&num_Z_planes, &max_num_Z_planes, 1, MPI_INT, MPI_MAX,
+                  _geometry->getMPICart());
+  }
+#endif
+  log_printf(NORMAL, "Max / total number of unique Z-planes %d / %d",
+             max_num_Z_planes, _num_Z_planes + 1);
```
Comment (GiudGiud) [inline src/TrackGenerator3D.cpp]: The total doesn't work that way. It'd be more code that it's worth to report it properly. I think the max is still somewhat relevant, the log could stay as INFO. What do you think ?

Comment (wuwenbin2006) [reply src/TrackGenerator3D.cpp]: Any necessity for this output? And I don't quite understand what is the meaning of this code snap

File: src/Universe.cpp (hunk 3 above)
Comment (wuwenbin2006) [reply src/Universe.cpp]: Good point

File: profile/Makefile (hunk 4 above)
Comment (GiudGiud) [reply profile/Makefile]: Somewhat. The code can be fit for SIMD vectorization but not vectorized by the compiler, if the compiler doesn't realize it's legal to SIMD vectorize at that point

File: src/CPUSolver.cpp [hunk 6]
```diff
@@ -147,9 +150,10 @@ void CPUSolver::setNumThreads(int num_threads) {
     if ((_track_generator->getSegmentFormation() == OTF_STACKS ||
        _track_generator->getSegmentFormation() == OTF_TRACKS) &&
        _track_generator->getNumThreads() != num_threads)
-      log_printf(WARNING, "The number of threads used in track generation "
-               "should match the number of threads used in the solver for OTF "
-               "ray-tracing methods, as threaded buffers are shared.");
+      log_printf(WARNING, "The number of threads used in track generation (%d)"
+               "should match the number of threads used in the solver (%d) for OTF "
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply src/CPUSolver.cpp]: fixing

File: src/TrackGenerator3D.cpp (hunk 5 above)
Comment (GiudGiud) [reply src/TrackGenerator3D.cpp]: The total number of Z-planes is somewhat interesting to the user. The fewer Z-planes the better for speed. 
Deleting this code because it doesn't provide that.

**Comment by GiudGiud** [comment]
Thanks for the review @wuwenbin2006 !
All your comments should be addressed now
